api/src/modules/impact/actual-vs-scenario.service.ts

```diff
@@ -0,0 +1,627 @@
+import { MaterialsService } from 'modules/materials/materials.service';
+import { GROUP_BY_VALUES } from 'modules/h3-data/dto/get-impact-map.dto';
+import { ImpactTableEntityType } from 'types/impact-table-entity.type';
+import { DEFAULT_PAGINATION, FetchSpecification } from 'nestjs-base-service';
+import { PaginatedEntitiesDto } from 'modules/impact/dto/paginated-entities.dto';
+import { GetMaterialTreeWithOptionsDto } from 'modules/materials/dto/get-material-tree-with-options.dto';
+import { LOCATION_TYPES } from 'modules/sourcing-locations/sourcing-location.entity';
+import { SOURCING_LOCATION_TYPE_BY_INTERVENTION } from 'modules/sourcing-locations/sourcing-location.entity';
+import { PaginationMeta } from 'utils/app-base.service';
+
+@Injectable()
+export class ActualVsScenarioImpactService {
+  //TODO: Hack to set a expected growing rate. This needs to be stored in the DB in the future
+  growthRate: number = 1.5;
+  logger: Logger = new Logger(ActualVsScenarioImpactService.name);
+
+  constructor(
+    private readonly indicatorService: IndicatorsService,
+    private readonly businessUnitsService: BusinessUnitsService,
+    private readonly adminRegionsService: AdminRegionsService,
+    private readonly suppliersService: SuppliersService,
+    private readonly materialsService: MaterialsService,
+    private readonly sourcingRecordService: SourcingRecordsService,
+  ) {}
+
+  async getActualVsScenarioImpactTable(
+    actualVsScenarioImpactTableDto: GetActualVsScenarioImpactTabledto,
+    fetchSpecification: FetchSpecification,
+  ): Promise<PaginatedImpactTable> {
+    const indicators: Indicator[] =
+      await this.indicatorService.getIndicatorsById(
+        actualVsScenarioImpactTableDto.indicatorIds,
+      );
+    this.logger.log('Retrieving data from DB to build Impact Table...');
+
+    //Getting Descendants Ids for the filters, in case Parent Ids were received
+    await this.loadDescendantEntityIds(actualVsScenarioImpactTableDto);
+
+    // Get full entity tree in cate ids are not passed, otherwise get trees based on
+    // given ids and add children and parent ids to them to get full data for aggregations
+    // TODO check if tree ids search is redundant
+    const entities: ImpactTableEntityType[] =
+      await this.getScenarioAndActualEntityTrees(
+        actualVsScenarioImpactTableDto,
+      );
+
+    const paginatedEntities: PaginatedEntitiesDto =
+      ActualVsScenarioImpactService.paginateRootEntities(
+        entities,
+        fetchSpecification,
+      );
+
+    this.updateGroupByCriteriaFromEntityTree(
+      actualVsScenarioImpactTableDto,
+      paginatedEntities.entities,
+    );
+
+    const dataForActualVsScenarioImpactTable: ActualVsScenarioImpactTableData[] =
+      await this.getDataForActualVsScenarioImpactTable(
+        actualVsScenarioImpactTableDto,
+        paginatedEntities.entities,
+      );
+
+    const processedDataForComparison: ActualVsScenarioImpactTableData[] =
```

Comment:
and here this method transforms ImpactTableData[] to ActualVsScenarioImpactTableData[]